### ISSUE TITLE
Design question Reactive state class with mutable values / `default_factory` for Reactive classes

### ISSUE THREAD

**Issue by Jhsmit** [widgetti/solara#58]
Suppose I'm making a solara component with several inputs/widgets,  with a corresponding `State` class, and assuming the return of this component is a list of ints. 

I would probably want to do something like this: 

```
@dataclasses.dataclass
class State:
    my_reactive_numbers = Reactive[List[int]]([])

```

And in my app create an instance of `State` and pass to the component. This works fine, however if i create multiple pairs of the component/state there could be a problem because of the mutable default argument passed to `Reactive` (the list), as shown here:

```python
s1 = State()
s2 = State()

s1.my_reactive_numbers.value = [1,2,3]
s2.my_reactive_numbers.value

>>> [1,2,3]

```

For non-mutable defaults such as `str`/`int`/`float` its not a problem. 

A solution could be to instead to this:

```python

class InitState():

    def __init__(self):
        self.my_reactive_numbers = Reactive([])
```

However, I would like to keep the possibility of using dataclasses. Perhaps `Reactive` variables should either optionally accept a `default_factory` kwarg (such as used for dataclasses; `dataclasses.field(default_factory=list)` ) or we should create a constructor which allows for passing of factory methods?

For example, in `class Reactive`:

```python
    @classmethod
    def from_factory(cls, factory: Callable[[], S]) -> "Reactive[S]":
        return cls(factory())
```

Then you could define your `State` class as:

```python
@dataclasses.dataclass
class State:
    my_reactive_numbers = Reactive[List[int]]([])

    my_reactive_factory_numbers = Reactive[List[int]].from_factory(list)

s1 = State()
s2 = State()

s1.my_reactive_factory_numbers = [4,5,6]
s2.my_reactive_factory_numbers.value
>>> []
```

**Comment by maartenbreddels** [comment]
First, dataclasses will not add a field if you do not do type annotations, so in your above State class, `s1.my_reactive_numbers is s2.my_reactive_numbers`. 

I am unsure if Reactive variables should go into dataclasses actually, since they aren't just 'data'. 
What do you think the advantage or reason is to do this?

**Comment by Jhsmit** [comment]
Yes I agree they aren't really data, but main reason I like to use them are is because 

1) their class attributes arent really class attributes and you get a free `__init__` from them.
2) You get a free `__repr__` (yes I know minor thing, but still)
And some other basic dataclass functionality, not directly sure what you need it for but 3) you get a free `__eq__`, 4) if you have immutables only, you can set `unsafe_hash=True` and use them as keys in a dict, etc

As you mentioned then to do things properly as a dataclass then a `factory` method might make more sense:

```python
    @classmethod
    def factory(cls, factory: Callable[[], S]) -> Callable[[], "Reactive[S]"]:
        def _factory():
            return cls(factory())

        return _factory
```
(the typing is from copilot)
And then create your state class as:

```python
@dataclasses.dataclass
class StateFactory:

    my_numbers: List[int] = dataclasses.field(default_factory=list)

    my_reactive_numbers: Reactive[List[int]] = dataclasses.field(default_factory=Reactive.factory(list))
```

The equivalent if you write your own `__init__` would be:

```python

class InitState:

    def __init__(self,
                 my_numbers: Optional[list[int]] = None,
                 my_reactive_numbers: Optional[List[int]] = None
                 ):
        
        self.my_numbers = my_numbers or []
        self.my_reactive_numbers: Reactive[list[int]] = Reactive(my_reactive_numbers or [])
```

Which is very cumbersome :)

If I would use the scatter example pattern where class attributes are used, I don't know how you would have multiple distinct copies of the state in one app, as they are shared between instances, even if you use a factory methods or immutables. 

```python

class StateCls:

    my_numbers: List[int] = dataclasses.field(default_factory=list)

    my_reactive_numbers = Reactive[List[int]]([])

    my_reactive_factory_numbers = Reactive[List[int]].from_factory(list)

    my_reactive_string = Reactive[str]("")
```

```python
sc1 = StateCls()
sc2 = StateCls()
sc1.my_reactive_numbers.value = [1,2,3]
sc2.my_reactive_numbers.value
>>> [1,2,3]

sc1.my_reactive_factory_numbers.value = [4,5,6]
sc2.my_reactive_factory_numbers.value
>>> [4,5,6]

sc1.my_reactive_string.value = 'foobar'
sc2.my_reactive_string.value
>>> 'foobar'

```

**Comment by Jhsmit** [comment]
For example: https://github.com/Jhsmit/awesome-solara/blob/master/examples/listitems_globalstate.py

**Comment by maartenbreddels** [comment]
> I don't know how you would have multiple distinct copies of the state in one app

The idea of the reactive variables, are that they are for application state, and you generally only need 1 of them (by definition).
If you need a reusable component, you shouldn't use global application state. 
To still be able to use the reactive variables, we'll provide a `use_reactive` hook (very soon!), so you don't need to fall back to `use_state.
Does that make sense?

Would it help to port the TODO app at https://reacton.solara.dev/en/latest/examples/ to use the Reactive variables? It requires this kind of 'nested' data.
